File: Solutions/677. Map Sum Pairs/677. Map Sum Pairs.cpp
//trie struct solution
struct trie{
    int sum=0;
    trie *ch[26]={};
};
class MapSum {
public:
    struct trie root;
    unordered_map<string,int> m;
    
    /** Initialize your data structure here. */
    MapSum() {
        
    }
    
    void insert(string key, int val) {
        auto p=&root;
        for(const auto& c:key){
            if(p->ch[c-'a']!=nullptr){
                p->ch[c-'a']=p->ch[c-'a'];
                p=p->ch[c-'a'];
            }else{
                p->ch[c-'a']=new trie();
                p=p->ch[c-'a'];
            }
            p->sum+=val-m[key];
        }
        m[key]=val;
    }
    
    int sum(string prefix) {
        auto p=&root;
        for(const auto& c:prefix){
            if(!p)
                break;
            p=p->ch[c-'a'];
        }
        return p?p->sum:0;
    }
};


//map solution
class MapSum {
public:
    map<string,int> m;
    /** Initialize your data structure here. */    
    void insert(string key, int val) {
        m[key]=val;
    }

    int sum(string prefix) {
        int sum=0,n=prefix.size();
        // iterator first points to "ap" in map 
        // n=="ap".size()==2
        // "ap".substr(0,n)==prefix("ap")
        for(auto it=m.lower_bound(prefix);it!=m.end() and it->first.substr(0, n)==prefix;++it){
            cout <<it->first <<"\n";
            sum+=it->second;
        }
        return sum;
    }
};

/**
 * Your MapSum object will be instantiated and called as such:
 * MapSum* obj = new MapSum();
 * obj->insert(key,val);
 * int param_2 = obj->sum(prefix);
 */

File: Solutions/677. Map Sum Pairs/677. Map Sum Pairs.go
type Trie struct {
    sum int
    ch [26]*Trie
}

type MapSum struct {
    root *Trie
    m map[string]int
}


/** Initialize your data structure here. */
func Constructor() MapSum {
    return MapSum{
        root: new(Trie),
        m: make(map[string]int),
    }
}

func (this *MapSum) Insert(key string, val int)  {
    p:=this.root
    m:=this.m
    for _,c:=range(key){
        if(p.ch[c-'a']!=nil){
            p.ch[c-'a']=p.ch[c-'a']
            p=p.ch[c-'a']
        }else{
            p.ch[c-'a']=new(Trie)
            p=p.ch[c-'a']
        }
        p.sum+=val-m[key]
    }
    m[key]=val
}


func (this *MapSum) Sum(prefix string) int {
    p:=this.root
    for i:=0;i<len(prefix)&&p!=nil;i+=1{
        p=p.ch[prefix[i]-'a']
    }
    if p!=nil{
        return p.sum
    }else{
        return 0
    }
}


/**
 * Your MapSum object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(key,val);
 * param_2 := obj.Sum(prefix);
 */